optimum/onnxruntime/io_binding/io_binding_helper.py

```diff
@@ -0,0 +1,170 @@
+        self.model_input_names = list(self.model_inputs.keys())
+        self.model_output_names = list(self.model_outputs.keys())
+
+    def prepare_io_binding(self, **kwargs):
+        """Returnas IO binding object for a session."""
+
+        name_to_np_type = TypeHelper.get_io_numpy_type_map(self.model)
+
+        # Bind inputs and outputs to onnxruntime session
+        io_binding = self.model.io_binding()
+
+        # Bind inputs
+        for input_name in self.model_input_names:
+            onnx_input = kwargs.pop(input_name)
+
+            assert onnx_input.is_contiguous()
+            io_binding.bind_input(
+                input_name,
+                onnx_input.device.type,
+                self.device.index,
+                name_to_np_type[input_name],
+                list(onnx_input.size()),
+                onnx_input.data_ptr(),
+            )
+
+        # Bind outputs
+        for name in self.model_output_names:
+            io_binding.bind_output(name, self.device.type, device_id=self.device.index)
+
+        return io_binding
+
+    @staticmethod
+    def to_pytorch(ort_value: OrtValue) -> torch.Tensor:
+        """Converts tensors held by OrtValues to torch tensor."""
+        env = {pkg.key for pkg in pkg_resources.working_set}
+        if "onnxruntime-training" in env:
+            return IOBindingHelper.to_pytorch_via_dlpack(ort_value)
+        else:
+            try:
+                return IOBindingHelper.to_pytorch_via_cupy(ort_value)
+            except Exception as e:
+                logging.error(traceback.format_exc())
+                logging.info("Unable to access output memory in CUDA, will offload to CPU")
+                return IOBindingHelper.to_pytorch_via_np(ort_value)
+
+    @staticmethod
+    def to_pytorch_via_np(ort_value: OrtValue) -> torch.Tensor:
+        ort_device = ort_value.device_name().lower()
+        return torch.tensor(ort_value.numpy()).to(ort_device)
+
+    @staticmethod
+    def to_pytorch_via_cupy(ort_value: OrtValue) -> torch.Tensor:
+        ort_device = ort_value.device_name().lower()
+        assert ort_device == "cuda", f"Convert via CuPy only when device is CUDA, got: {ort_device}"
+
+        ort_type = ort_value.data_type()
+        np_type = TypeHelper.ort_type_to_numpy_type(ort_type)
+
+        # Access CUDA memory via CuPy
+        import cupy as cp
+
+        memory = cp.cuda.UnownedMemory(ort_value.data_ptr(), 0, None)
+        memory_ptr = cp.cuda.MemoryPointer(memory, 0)
+        cp_array = cp.ndarray(shape=ort_value.shape(), memptr=memory_ptr, dtype=np_type)
+        torch_tensor = torch.from_dlpack(cp_array.toDlpack())
+
+        # If is boolean, the dtype will be uint8 and need to be convert back to bool.
+        if "bool" in ort_type:
+            torch_tensor = torch_tensor.to(torch.bool)
+
+        torch_tensor = torch_tensor.clone()
+
+        return torch_tensor
+
+    @staticmethod
+    # only `onnxruntime-training` supports dlpack for OrtValue
```

Comment:
Not sure I understand the implication of this comment. Could you expand it a bit?

Reply:
So are the remaining `is_cupy_available` and `is_onnxruntime_training_available` methods useful here (maybe a future PR)?

Reply:
@fxmarty, oh yeah, I will move them to another PR. Good catch!